src/Analysis/Ast/Impl/Analyzer/PythonAnalyzerSession.cs

```diff
@@ -147,30 +146,26 @@ internal sealed class PythonAnalyzerSession {
             try {
                 _log?.Log(TraceEventType.Verbose, $"Analysis version {Version} of {originalRemaining} entries has started.");
                 remaining = await AnalyzeAffectedEntriesAsync(stopWatch);
+                Debug.Assert(_ace.Count == 0);
             } finally {
                 stopWatch.Stop();
 
-                bool isCanceled;
-                bool isFinal;
+                var isFinal = false;
                 lock (_syncObj) {
-                    isCanceled = _isCanceled;
+                    if (!_isCanceled) {
+                        _progress.ReportRemaining(remaining);
+                    }
+
                     _state = State.Completed;
-                    isFinal = _walker.MissingKeys.Count == 0 && !isCanceled && remaining == 0;
+                    isFinal = _walker.MissingKeys.Count == 0 && !_isCanceled && remaining == 0;
                     _walker = null;
                 }
 
-                if (!isCanceled) {
-                    _progress.ReportRemaining(remaining);
-                    if (isFinal) {
-                        var (modulesCount, totalMilliseconds) = ActivityTracker.EndTracking();
-                        totalMilliseconds = Math.Round(totalMilliseconds, 2);
-                        (_analyzer as PythonAnalyzer)?.RaiseAnalysisComplete(modulesCount, totalMilliseconds);
-                        _log?.Log(TraceEventType.Verbose, $"Analysis complete: {modulesCount} modules in {totalMilliseconds} ms.");
-                        //#if DEBUG
-                        //                        var notReady = _analyzer.LoadedModules.Where(m => (m.ModuleType == ModuleType.Library || m.ModuleType == ModuleType.Stub) && m.Analysis is EmptyAnalysis).ToArray();
-                        //                        Debug.Assert(notReady.Length == 0);
-                        //#endif
-                    }
+                if (isFinal) {
+                    var (modulesCount, totalMilliseconds) = ActivityTracker.EndTracking();
+                    totalMilliseconds = Math.Round(totalMilliseconds, 2);
+                    _analyzer.RaiseAnalysisComplete(modulesCount, totalMilliseconds);
+                    _log?.Log(TraceEventType.Verbose, $"Analysis complete: {modulesCount} modules in {totalMilliseconds} ms.");
```

Comment:
One thing I'm seeing here is that due to the lazy loading, we're (in a way) under-reporting the number of modules we're working with. Hopefully that's not such a big deal if the work also disappears, but it may end up appearing as though we're not "improving" per-module because for the N modules we still have to load, we have to do the same work.

Reply:
Not entirely happy with `RaiseAnalysisComplete` being bool, but I can make it return if it actually raised the event. Somewhat ugly, but implementation uses data internal to the analyzer which i'd rather not make public.

Reply:
I'm not sure what you mean by returning a bool, I'm more talking about the fact that we'll log "123 modules analyzed", when maybe there are a bunch more that weren't "loaded" because they'll be fetched from disk.

You don't have to change anything for this comment, I was more thinking about the impact of lazyness on our reporting.

Reply:
I mean returning bool from `RaiseAnalysisComplete` and then only reporting when it is really done.